blog/_src/posts/2019-08-19-scoping-in-r.md

```diff
@@ -0,0 +1,491 @@
+    expression `x <- 1` assigns the value `1` to the variable `x` in the
+    current environment. If a mapping for `x` already exists in the
+    environment, then the assignment will update and overwrite the existing
+    value. Otherwise, a new mapping is created in the environment. Note that
+    variable assignment can only update the current environment, and never
+    creates a scope.
+
+From this description, we can see that R implements lexical scoping (or at
+least, something that behaves a lot like lexical scoping): each function value
+is associated with the environment it was evaluated in, and variable lookup
+proceeds along the chain of enclosing environments. In fact, the creators of R
+have confirmed that lexical scoping was their intent.[^2]
+
+On the other hand, variable lookup depends on the run-time state of the
+program---names cannot be resolved statically. Furthermore, since R provides
+operations for environment manipulation, a programmer can easily circumvent
+lexical scoping.
+
+The following examples will make this clear.
+
+[^2]: R. Gentleman and R. Ihaka. "Lexical Scope and Statistical Computing",
+_Journal of Computational and Graphical Statistics_, vol. 9, no. 3, 2000.
+\[[DOI](https://doi.org/10.1080/10618600.2000.10474895)\]\[[Available
+online](https://www.stat.auckland.ac.nz/~ihaka/downloads/lexical.pdf)\]
+
+
+### Examples
+
+#### Adding variable mappings at run time
+
+```r
+x <- 1
+f <- function() {
+  g <- function() x
+  x <- 2
+  g()
+}
+f() # 2
+```
+
+When `f` is called, it creates a function `g` that returns `x`, assigns `2` to
+`x`, and then calls `g`. When `g` is called, it looks up `x`. Since no mapping
+is found in `g`'s environment, it searches in the enclosing environment
+(`f`'s), and finds that `x` has value `2`. Therefore, `g` returns `2`.
+
+Note that the `x` on line 3 is resolved only when function `g` is called, not
+when it is defined. However, when `g` is defined, its environment has a
+reference to `f`'s environment. Therefore, as long as `x` is defined _before_
+`g` is called, the lookup will always succeed.
+
+Here's a second example:
+
+```r
+x <- 1
+f <- function(b) {
+  if (b)
+    x <- 2
+  x
+}
+f(TRUE)  # 2
+f(FALSE) # 1
+```
+
+`f` is a function that branches on its argument, `b`. If `b` evaluates to true,
+then the expression `x <- 2` is evaluated, and a mapping for `x` is created in
+`f`'s environment. Otherwise, no mapping is created.
+
+When we look up the value of `x` on line 5, R will first search the function's
+environment. If `b` evaluated to true, then R will find a value for `x`, which
+is `2`. Otherwise, R will search in the enclosing environment of `f`, and find
+that `x` is `1`.
+
+Both of these examples vaguely resemble dynamic scoping, in that `x` takes the
+value of the most recent assignment. However, this is not how R is
+implemented---as there is no traversal of the call stack---and it is not
+consistent with how R behaves in other examples.
```

Comment:
The previous post doesn't say anything about traversing any call stack. Using those definitions, these examples are dynamic.

Maybe change to:
> Both of these example defy lexical scoping, in that ....
> However, the next section (link?) argues that R is lexically-scoped at heart.

Reply:
Yeah, I'll have to think about this a bit. You're totally right that I never mentioned the call stack in the previous post, and I do want to keep the definition general. But I also don't want to say this example demonstrates dynamic scope. The lookup mechanism is definitely lexical-like, and if you take my intuition that it's _environments_ and not _variables_ that are being referenced, then this is lexically scoped.

Reply:
```diff
 Both of these examples vaguely resemble dynamic scoping, in that `x` takes the
-value of the most recent assignment. However, this is not how R is
-implemented---as there is no traversal of the call stack---and it is not
-consistent with how R behaves in other examples.
+value of the most recent assignment. However, this is not how R is implemented,
+and it is not consistent with how R behaves in other examples.
```